### ISSUE TITLE
Updating Series.dims on Series.applyKeys

### ISSUE THREAD

**Issue by jwittenbach** [thunder-project/thunder#219]
Currently, calling `Series.applyKeys` returns a new `Series` copies the value of the `Series.dims` property from the original `Series`. However, a call to `Series.applyKeys` is very likely to change the range of the keys, so this may not be appropriate.

Example:
```
In [49]: data.dims
Out[49]: Dimensions(values=[(0, 0, 0), (75, 86, 1)], n=3)

In [50]: data.applyKeys(lambda k: tuple([i-1 for i in k])).dims
Out[50]: Dimensions(values=[(0, 0, 0), (75, 86, 1)], n=3)
```

**Comment by jwittenbach** [comment]
There is definitely a solid use-case for `Series` -- I've been dealing with data that gets loaded in with the keys starting at `1` and I need to reindex them to start at `0` for later use of `Series.pack`. There may be similar use-cases on `Images`, in which case we might want to take a look at implementing a fix for this on `Images` as well.

**Comment by freeman-lab** [comment]
I agree that your suggestion would solve the issue for now, but I'm generally in favor of reducing direct access to keys and values, especially with the transition to Bolt in mind, where direct access to keys should never be necessary as part of the public API, and shape information will be propagated correctly appropriately by other functions (like `map`).

This particular case might best be handled as an optional flag during loading, though I'm not sure that would handle every current use of `applyKeys`.

And note that in the case of images, the `dims` actually refers to the dimensions of the values, so the same issue arises for `applyValues`, but there's no metadata that's key-specific.

**Comment by jwittenbach** [comment]
That makes a lot of sense. The one thing I would add is that, at least for `Series`, access to the keys will still remain essential all the way up until Bolt has a sparse array, which could be a ways off. 

Should we try to implement something to fix this for now?
